enhancements/opm-package-filtering.md

```diff
@@ -0,0 +1,79 @@
+---
+title: opm-package-filtering
+authors:
+  - "@djzager"
+reviewers:
+  - TBD
+approvers:
+  - TBD
+creation-date: 2020-04-03
+last-updated: 2020-04-03
+status: implementable
+---
+
+# OPM Package Filtering
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+
+## Summary
+
+Add support to the `opm` command line utility to filter operators from an
+operator index (or registry). This enables consumers of operator indexes to take
+an existing, and presumably large, index and generate a new index with only the
+operators they wish to make available in the catalog. In short, this is the
+inverse of `opm (index|registry) rm` where you select which operators to remove,
+`opm (index|registry) filter` allows you to select the operators to keep.
+
+## Motivation
+
+As Operator adoption expands and OLM usage increases the index images used to
+make operator catalogs available to OLM will grow. Cluster administrators
+__should__ have a way to curate operator catalogs. This feature provides a basic
+mechanism for catalog curation.
+
+### Goals
+
+- Allow `opm` user to select operators to keep from an existing registry or
+  index image.
+
+### Non-Goals
+
+- Manipulate the available channels of any operator in the registry or index
+  image.
+
+## Proposal
+
+### Filter Subcommand
+
+Adding the `filter` subcommand to the `index` and `registry` command provides a
+simple mechanism for selecting operators to keep from an index image or a
+registry database (depending on the root command).
+
+### User Story
+
+#### Pruning an Existing Operator Catalog
+
+As a cluster administrator, Billy knows what operators from an existing operator
+index he wants to make available to his cluster's users. Billy can build a new
+operator index by selecting operators from the previous index to keep.
+
+### Implementation Details
+
+#### New Commands
+
+A `filter` subcommand will be added to both the `index` and `registry` command
+mimmicking the behavior of `rm` in all but one significant way. The `operators`
```

Comment:
if it's a new subcommand, could it not use `packages` in both instances? Alternatively, the package list could be the positional parameters of the command, rather than a named parameter.

Reply:
My only struggle with this idea is wanting consistency. For example, when you remove operators from an index by saying `opm index rm --from-index <index-img> --operators <op1>,<op2>` and it seems fair for a user to think prune would look something like `opm index prune --from-index <index-img> --operators <op1>,<op2>`.

It may be that `packages` is really most appropriate for all of `opm (index|registry) (rm|prune)` but I'm hesitant to do that for all of `opm (index|registry) prune` without consensus. WDYT @ecordell ?

Reply:
I like being consistent, though IMO the existing commands are probably not following typical UX conventions; e.g no primary positional argument, required options, etc.